sherpa-onnx/csrc/context-graph.h

```diff
@@ -0,0 +1,64 @@
+// sherpa-onnx/csrc/context-graph.h
+//
+// Copyright (c)  2023  Xiaomi Corporation
+
+#ifndef SHERPA_ONNX_CSRC_CONTEXT_GRAPH_H_
+#define SHERPA_ONNX_CSRC_CONTEXT_GRAPH_H_
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace sherpa_onnx {
+
+class ContextState;
+class ContextGraph;
+using ContextStatePtr = std::shared_ptr<ContextState>;
+using ContextGraphPtr = std::shared_ptr<ContextGraph>;
+
+struct ContextState {
+  int32_t token;
+  float token_score;
+  float node_score;
+  float local_node_score;
+  bool is_end;
+  std::unordered_map<int32_t, ContextStatePtr> next;
+  ContextStatePtr fail;
+  ContextStatePtr output;
+
+  ContextState() = default;
+  ContextState(int32_t token, float token_score, float node_score,
+               float local_node_score, bool is_end)
+      : token(token),
+        token_score(token_score),
+        node_score(node_score),
+        local_node_score(local_node_score),
+        is_end(is_end) {}
+};
+
+class ContextGraph {
+ public:
+  ContextGraph() = default;
+  explicit ContextGraph(float context_score) : context_score_(context_score) {
+    root_ = std::make_shared<ContextState>(-1, 0, 0, 0, false);
+    root_->fail = root_;
```

Comment:
Don't want to use the raw pointer, I use a weak_ptr to fix this.

Reply:
I suggest that you use `std::unique_ptr<>`.

Here are the reasons:

The pointers in the map `next`  are allocated in the `Build()` method. Once allocated they will not be deallocated as long as `root_` is alive. 

So `std::unique_ptr<>` is a perfect choice for it. It is smaller in size than `std::shared_ptr<>`. Also, when you copy a shared pointer in `ForwardOneStep()` and `FillFailOutput()`, you need to pay for increasing/decreasing an atomic counter.

Also, you don't need to use `weak_ptr` once switching to `unique_ptr`.


Is there a reason to choose `std::shared_ptr` instead of `std::unique_ptr`?

If you want to return a ContextState to a stream, you only need to return the underlying raw pointer managed by the unique ptr.
Since the `root_` is kept as a member of the stream, so you don't need to worry about dangling pointers.